typescript/agentkit/src/action-providers/compound/compoundActionProvider.ts

```diff
@@ -0,0 +1,419 @@
+
+      // Check that there is enough collateral supplied to withdraw
+      const collateralBalance = await getCollateralBalance(wallet, cometAddress, tokenAddress);
+      if (amountAtomic > collateralBalance) {
+        const humanBalance = formatAmountFromDecimals(collateralBalance, decimals);
+        return `Error: Insufficient balance. Trying to withdraw ${args.amount}, but only have ${humanBalance} supplied`;
+      }
+
+      // Check if position would be healthy after withdrawal
+      const projectedHealthRatio = await getHealthRatioAfterWithdraw(
+        wallet,
+        cometAddress,
+        tokenAddress,
+        amountAtomic,
+      );
+      if (projectedHealthRatio.lessThan(1)) {
+        return `Error: Withdrawing ${args.amount} would result in an unhealthy position. Health ratio would be ${projectedHealthRatio.toFixed(2)}`;
+      }
+
+      // Withdraw from Compound
+      const data = encodeFunctionData({
+        abi: COMET_ABI,
+        functionName: "withdraw",
+        args: [tokenAddress, amountAtomic],
+      });
+
+      const txHash = await wallet.sendTransaction({
+        to: cometAddress,
+        data,
+      });
+      await wallet.waitForTransactionReceipt(txHash);
+
+      // Get current and new health ratios and token symbol
+      const currentHealth = await getHealthRatio(wallet, cometAddress);
+      const newHealth = await getHealthRatio(wallet, cometAddress);
+      const tokenSymbol = await getTokenSymbol(wallet, tokenAddress);
+
+      return (
+        `Withdrawn ${args.amount} ${tokenSymbol} from Compound.\n` +
+        `Transaction hash: ${txHash}\n` +
+        `Health ratio changed from ${currentHealth.toFixed(2)} to ${newHealth.toFixed(2)}`
+      );
+    } catch (error) {
+      return `Error withdrawing from Compound: ${error instanceof Error ? error : error && typeof error === "object" && "message" in error ? `Error: ${error.message}` : error}`;
+    }
+  }
+
+  /**
+   * Borrows base assets from Compound.
+   *
+   * @param wallet - The wallet instance to perform the transaction.
+   * @param args - The input arguments including assetId and amount.
+   * @returns A message indicating success or an error message.
+   */
+  @CreateAction({
+    name: "borrow",
+    description: `
+This tool allows borrowing base assets from Compound.
+It takes:
+- assetId: The asset to borrow, either 'weth' or 'usdc'
+- amount: The amount of base tokens to borrow in human-readable format
+Examples:
+- 1000 USDC
+- 0.5 WETH
+Important notes:
+- Use the exact amount provided
+- Ensure you have sufficient collateral to borrow
+    `,
+    schema: CompoundBorrowSchema,
+  })
+  async borrow(
+    wallet: EvmWalletProvider,
+    args: z.infer<typeof CompoundBorrowSchema>,
+  ): Promise<string> {
+    try {
+      const cometAddress = getCometAddress(wallet.getNetwork());
+
+      // USDC has 6 decimals
+      const amountAtomic = formatAmountWithDecimals(args.amount, 6);
```

Comment:
What if the borrow input is WETH? This might be relevant: https://github.com/coinbase/agentkit/blob/main/typescript/agentkit/src/action-providers/erc20/erc20ActionProvider.ts#L47-L52